docs/architecture/adr-009-transport-agnostic-peer-abstraction.md

```diff
@@ -0,0 +1,194 @@
+# ADR 009: Transport agnostic Peer abstraction
+
+## Changelog
+* 2021-02-05: drafted
+
+## Context
+
+With the opportunity to design and implement the peer-to-peer stack from
+scratch in the context of the Tendermint implementation in Rust, a lot of the
+learnings of the shortcomings of the original Go implementation can be used to
+prevent certain mistakes. Namely two:
+
+* Leakage of physical concerns into the core domain
+* Flexibility to adopt different wire protocols for transport of messages
+
+For that, the first set of newly introduced concepts will attempt to be generic
+over which transport is used to connect and converse with other peers. Given
+strongly tailored abstract interfaces, concrete implementations will be easy to
+spin up and plug into the machinery which lifts bytes from the wire into the
+core domain and transports messages into the rest of the system.
+
+## Decision
+
+### Transport
+
+Wraping the design is the `Transport`. Modelled with the properties of
+a physical network endpoint in mind, which can be bound and stopped. It should
+strongly correspond to the acquisition and lifecycle management of network
+resources on the system.
+
+``` rust
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;
+
+    fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;
+    fn shutdown(&self) -> Result<()>;
+}
+```
+
+After the successful bind the caller holds an `Endpoint` as well as a stream of
+incoming `Connection`s. Which is a standardised way to connect to new peers and
+react to newly connected ones respectively.
+
+``` rust
+pub trait Endpoint: Send {
+    type Connection;
+
+    fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;
+    fn listen_addrs(&self) -> Vec<SocketAddr>;
+}
+```
+
+Centerpiece of the whole shebang is the `Connection`. It represents a connected
+peer and provides the primitives to get data and send data from a peer. It is
+designed with the outlook to support stream based transports down the road.
+While being open to enable feature parity with current production installations
+based on tendermint-go's `MConn`.
+
+``` rust
+pub trait Connection: Send {
+    type Error: std::error::Error + Send + Sync + 'static;
+    type Read: Read;
+    type Write: Write;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr>;
+    fn close(&self) -> Result<()>;
+    fn local_addr(&self) -> SocketAddr;
+    fn open_bidirectional(
+        &self,
+        stream_id: &StreamId,
+    ) -> Result<(Self::Read, Self::Write), Self::Error>;
+    fn public_key(&self) -> PublicKey;
+    fn remote_addr(&self) -> SocketAddr;
```

Comment:
For the closed case we should be able to return the information from when the connection was running. Is there ever the case where we have a connection that is not open yet?

Reply:
Right, I was just looking at `TcpStream` `local_addr` and `peer_addr` methods that both return `Result<SocketAddr>` (rather than `SocketAddr`) even though it's also not possible to construct a non-connected stream. 

At the moment, this is how `local_addr` implemented:

```rust
    fn local_addr(&self) -> SocketAddr {
        self.stream
            .local_addr()
            .unwrap()
    }
```

if there's an error, the program will panic, which is not good.

Reply:
It doesn't need to be done that way, right? We can also memorise this information when we construct the connection and return the stored value whenever the method is called. Or am I missing an important detail here? In my understanding the address shouldn't change over the lifetime of the conn.